crates/rattler_installs_packages/src/index/package_sources.rs

```diff
@@ -0,0 +1,108 @@
+use crate::types::NormalizedPackageName;
+use std::collections::BTreeMap;
+use url::Url;
+
+struct PackageSource {
+    alias: String,
+    url: Url,
+}
+
+/// "Builder" pattern for creating a [`PackageSources`] instance
+pub struct PackageSourcesBuilder {
+    base_source: Url,
+    extra_sources: Vec<PackageSource>,
+    overrides: BTreeMap<NormalizedPackageName, String>,
+}
+
+impl PackageSourcesBuilder {
+    /// Start building a new [`PackageSources`] instance, with the given base index URL
+    /// This URL will be used for all packages by default
+    pub fn new(base_index_url: Url) -> Self {
+        Self {
+            base_source: base_index_url,
+            extra_sources: Default::default(),
+            overrides: Default::default(),
+        }
+    }
+
+    /// Add another index URL
+    pub fn with_index(&mut self, alias: String, url: Url) -> &mut Self {
```

Comment:
I always prefer to use `mut self` and return `Self` for functions called `with_*`. This allows chaining in such a way that returns a value in the end.

If you want functions that take `&mut self` and return `&mut Self` consider naming them `add_*` or `set_*`. 

For builders I often add both variants.